design-documents/web-api-controllers.md

```diff
@@ -0,0 +1,374 @@
+ 
+ 
+## GraphQL
+#### Current situation
+GraphQL (sometimes) utilizes service contracts and relies on them to have execute complex validations if any needed.
+It also uses simple validation provided by graphql schema (like a property not allowed to be null). However if an
+endpoint requires complex validation/authorization that is not performed by a service contract (and it shouldn't since
+service contracts must not be aware of context) resolvers would duplicate to match functionality provided by regular
+controllers providing HTML presentation. Furthermore since it's easier to define simple validation rules via graphQL
+schema GraphQL gateway may provide more proper user input validation than HTML/REST/SOAP gateways.
+ 
+ 
+## HTML Presentation
+#### Current situation
+Controllers implementing _ActionInterface_ do complex validation/authorization themselves which is fine as they are
+supposed to be responsible for user input processing, still there's a problem with other areas not being able to reuse
+this logic.
+ 
+ 
+## What needs to be done to solve this?
+#### Using operation-specific DTOs
+In order to make it more clear what data will be actually used/allowed for a certain operation and to avoid transferring
+of extensive data sets we should move from having generic DTOs describing entities to operation-specific DTOs describing
+arguments and results of an operation.
+ 
+Example on how would we change DTOs for operations regarding customer creation/update:
+ 
+First we would have DTO that describes a customer in the way it's convenient for client code when updating a customer:
+```php
+interface CustomerUpdateInterface
+{
+    //No getId() since storefront clients must not be able to control their customer's ID
+    ...
+    
+    //For setting new password
+    public function getPassword(): ?string;
+    
+    public function getAddresses(): CustomerAddress[];
+}
+```
+ 
+Then we would have DTO for customer's addresses:
+```php
+interface CustomerAddressUpdateInterface
+{
+    public function getId(): ?string;
+    ....
+    
+    //No getCustomerId() because we don't give control over that
+    ....
+    
+    public function getPostcode(): string
+}
+```
+
+Since we don't want original password in customer responses we would not return it as read operations result:
+```php
+interface CustomerReadInterface
+{
+    //Now we have getId() since it's OK for storefront clients to know their customer's ID.
+    public function getId(): string;
+    
+    ....
+    //No getPassword()
+    ....
+}
+```
+ 
+By having these operation-specific DTOs we would only have the data we need as arguments and responses.
+ 
+ 
+#### Use declarative validation
+Web API clients displaying forms should be able to get a list of invalid properties sent to display to end users or
+even have validation rules provided to them. Right now entities in Magento are being validated manually and errors
```

Comment:
I don't touch that topic here and I'm not sure core Magento needs it but with declarative validation rules it will be possible for merchants to implement